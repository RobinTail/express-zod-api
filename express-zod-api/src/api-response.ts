import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
