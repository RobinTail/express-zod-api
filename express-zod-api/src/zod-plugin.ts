/**
 * @fileoverview Zod Runtime Plugin
 * @see https://github.com/colinhacks/zod/blob/90efe7fa6135119224412c7081bd12ef0bccef26/plugin/effect/src/index.ts#L21-L31
 * @desc This code modifies and extends zod's functionality immediately when importing express-zod-api
 * @desc Enables .example() and .deprecated() on all schemas (ZodType)
 * @desc Enables .label() on ZodDefault
 * @desc Enables .remap() on ZodObject
 * @desc Stores the argument supplied to .brand() on all schema (runtime distinguishable branded types)
 * */
import * as R from "ramda";
import { z } from "zod/v4";
import { getExamples, metaSymbol } from "./metadata";
import { Intact, Remap } from "./mapping-helpers";
import type {
  $ZodType,
  $ZodShape,
  $ZodLooseShape,
  $ZodObjectConfig,
  $ZodCheck,
  $ZodCheckInternals,
  $ZodCheckDef,
  SomeType,
  $ZodDefaultInternals,
  $ZodDefault,
  $ZodObjectInternals,
  $ZodObject,
  $ZodTypeInternals,
  $strip,
} from "zod/v4/core";

declare module "zod/v4/core" {
  interface GlobalMeta {
    default?: unknown; // can be an actual value or a label like "Today"
    examples?:
      | unknown[] // see zod commit ee5615d
      | Record<string, { value: unknown; [k: string]: unknown }>; // @todo remove in v25
    /** @deprecated use examples instead */
    example?: unknown; // see zod commit ee5615d @todo remove in v25
  }
}

declare module "zod/v4" {
  interface ZodType<
    out Output = unknown,
    out Input = unknown,
    out Internals extends $ZodTypeInternals<Output, Input> = $ZodTypeInternals<
      Output,
      Input
    >,
  > extends $ZodType<Output, Input, Internals> {
    /** @desc Alias for .meta({examples}), but argument is typed to ensure the correct placement for transformations */
    example(example: z.output<this>): this;
    deprecated(): this;
  }
  interface ZodDefault<T extends SomeType = $ZodType>
    extends z._ZodType<$ZodDefaultInternals<T>>,
      $ZodDefault<T> {
    /** @desc Change the default value in the generated Documentation to a label, alias for .meta({ default }) */
    label(label: string): this;
  }
  interface ZodObject<
    // @ts-expect-error -- external issue
    out Shape extends $ZodShape = $ZodLooseShape,
    out Config extends $ZodObjectConfig = $strip,
  > extends z._ZodType<$ZodObjectInternals<Shape, Config>>,
      $ZodObject<Shape, Config> {
    remap<V extends string, U extends { [P in keyof Shape]?: V }>(
      mapping: U,
    ): z.ZodPipe<
      z.ZodPipe<this, z.ZodTransform>, // internal type simplified
      z.ZodObject<Remap<Shape, U, V> & Intact<Shape, U>, Config>
    >;
    remap<U extends $ZodShape>(
      mapper: (subject: Shape) => U,
    ): z.ZodPipe<z.ZodPipe<this, z.ZodTransform>, z.ZodObject<U>>; // internal type simplified
  }
}

interface $EZBrandCheckDef extends $ZodCheckDef {
  check: "$EZBrandCheck";
  brand?: string | number | symbol;
}

interface $EZBrandCheckInternals extends $ZodCheckInternals<unknown> {
  def: $EZBrandCheckDef;
}

interface $EZBrandCheck extends $ZodCheck {
  _zod: $EZBrandCheckInternals;
}

/**
 * This approach was suggested to me by Colin in a PM on Twitter.
 * Refrained from storing the brand in Metadata because it should withstand refinements.
 * */
const $EZBrandCheck = z.core.$constructor<$EZBrandCheck>(
  "$EZBrandCheck",
  (inst, def) => {
    z.core.$ZodCheck.init(inst, def);
    inst._zod.onattach.push((schema) => (schema._zod.bag.brand = def.brand));
    inst._zod.check = () => {};
  },
);

const exampleSetter = function (this: z.ZodType, value: z.output<typeof this>) {
  const examples = getExamples(this).slice();
  examples.push(value);
  return this.meta({ examples });
};

const deprecationSetter = function (this: z.ZodType) {
  return this.meta({ deprecated: true });
};

const labelSetter = function (this: z.ZodDefault, defaultLabel: string) {
  return this.meta({ default: defaultLabel });
};

const brandSetter = function (
  this: z.ZodType,
  brand?: string | number | symbol,
) {
  return this.check(new $EZBrandCheck({ brand, check: "$EZBrandCheck" }));
};

type _Mapper = <T extends Record<string, unknown>>(
  subject: T,
) => { [P in string | keyof T]: T[keyof T] };

const objectMapper = function (
  this: z.ZodObject,
  tool: Record<string, string> | _Mapper,
) {
  const transformer =
    typeof tool === "function" ? tool : R.renameKeys(R.reject(R.isNil, tool)); // rejecting undefined
  const nextShape = transformer(
    R.map(R.invoker(0, "clone"), this._zod.def.shape), // immutable, changed from R.clone due to failure
  );
  const hasPassThrough = this._zod.def.catchall instanceof z.ZodUnknown;
  const output = (hasPassThrough ? z.looseObject : z.object)(nextShape); // proxies unknown keys when set to "passthrough"
  return this.transform(transformer).pipe(output);
};

if (!(metaSymbol in globalThis)) {
  (globalThis as Record<symbol, unknown>)[metaSymbol] = true;
  for (const entry of Object.keys(z)) {
    if (!entry.startsWith("Zod")) continue;
    if (/(Success|Error|Function)$/.test(entry)) continue;
    const Cls = z[entry as keyof typeof z];
    if (typeof Cls !== "function") continue;
    Object.defineProperties(Cls.prototype, {
      ["example" satisfies keyof z.ZodType]: {
        value: exampleSetter,
        writable: false,
      },
      ["deprecated" satisfies keyof z.ZodType]: {
        value: deprecationSetter,
        writable: false,
      },
      ["brand" satisfies keyof z.ZodType]: {
        set() {}, // this is required to override the existing method
        get() {
          return brandSetter.bind(this) as z.ZodType["brand"];
        },
      },
    });
  }

  Object.defineProperty(
    z.ZodDefault.prototype,
    "label" satisfies keyof z.ZodDefault,
    { value: labelSetter, writable: false },
  );
  Object.defineProperty(
    z.ZodObject.prototype,
    "remap" satisfies keyof z.ZodObject,
    { value: objectMapper, writable: false },
  );
}
