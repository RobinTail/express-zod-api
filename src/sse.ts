import { Response } from "express";
import { z } from "zod";
import { EmptySchema, FlatObject } from "./common-helpers";
import { CommonConfig } from "./config-type";
import { contentTypes } from "./content-type";
import { EndpointsFactory } from "./endpoints-factory";
import { Middleware } from "./middleware";
import { ResultHandler } from "./result-handler";
import {
  ensureHttpError,
  getPublicErrorMessage,
  logServerError,
} from "./result-helpers";

type EventsMap = Record<string, z.ZodTypeAny>;

export interface Emitter<E extends EventsMap> extends FlatObject {
  /** @desc Returns true when the connection was closed or terminated */
  isClosed: () => boolean;
  /** @desc Sends an event to the stream accordin to the declared schema */
  emit: <K extends keyof E>(event: K, data: z.input<E[K]>) => void;
}

export const makeEventSchema = (event: string, data: z.ZodTypeAny) =>
  z.object({
    data,
    event: z.literal(event),
    id: z.string().optional(),
    retry: z.number().int().positive().optional(),
  });

export const formatEvent = <E extends EventsMap>(
  events: E,
  event: keyof E,
  data: unknown,
) =>
  makeEventSchema(String(event), events[event])
    .transform((props) =>
      [
        `event: ${props.event}`,
        `data: ${JSON.stringify(props.data)}`,
        "",
        "", // empty line: events separator
      ].join("\n"),
    )
    .parse({ event, data });

const headersTimeout = 1e4; // 10s to respond with a status code other than 200
export const ensureStream = (response: Response) =>
  response.headersSent ||
  response.writeHead(200, {
    connection: "keep-alive",
    "content-type": contentTypes.sse,
    "cache-control": "no-cache",
  });

export const makeMiddleware = <E extends EventsMap>(events: E) =>
  new Middleware({
    handler: async ({ response }): Promise<Emitter<E>> =>
      setTimeout(() => ensureStream(response), headersTimeout) && {
        isClosed: () => response.writableEnded || response.closed,
        emit: (event, data) => {
          ensureStream(response);
          response.write(formatEvent(events, event, data), "utf-8");
          response.flush();
        },
      },
  });

export const makeResultHandler = <E extends EventsMap>(events: E) =>
  new ResultHandler({
    positive: () => {
      const [first, ...rest] = Object.entries(events).map(([event, schema]) =>
        makeEventSchema(event, schema),
      );
      return {
        mimeType: contentTypes.sse,
        schema: rest.length
          ? z.discriminatedUnion("event", [first, ...rest])
          : first,
      };
    },
    negative: { mimeType: "text/plain", schema: z.string() },
    handler: async ({ response, error, logger, request, input }) => {
      if (error) {
        const httpError = ensureHttpError(error);
        logServerError(httpError, logger, request, input);
        if (!response.headersSent) {
          response
            .status(httpError.statusCode)
            .type("text/plain")
            .write(getPublicErrorMessage(httpError), "utf-8");
        }
      }
      response.end();
    },
  });

export class EventStreamFactory<
  E extends EventsMap,
  TAG extends string,
> extends EndpointsFactory<EmptySchema, Emitter<E>, string, TAG> {
  constructor({ events, config }: { events: E; config?: CommonConfig<TAG> }) {
    super({ config, resultHandler: makeResultHandler(events) });
    this.middlewares = [makeMiddleware(events)];
  }
}
