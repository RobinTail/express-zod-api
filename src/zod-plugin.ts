/**
 * @fileoverview Zod Runtime Plugin
 * @see https://github.com/colinhacks/zod/blob/90efe7fa6135119224412c7081bd12ef0bccef26/plugin/effect/src/index.ts#L21-L31
 * @desc This code modifies and extends zod's functionality immediately when importing express-zod-api
 * @desc Enables .examples() on all schemas (ZodType)
 * @desc Enables .label() on ZodDefault
 * @desc Stores the argument supplied to .brand() on all schema (runtime distinguishable branded types)
 * */
import { clone, fromPairs, map, pipe, toPairs, pair } from "ramda";
import { z } from "zod";
import { FlatObject } from "./common-helpers";
import { cloneSchema, Metadata, metaSymbol } from "./metadata";
import { Intact, Remap } from "./mapping-helpers";

declare module "zod" {
  interface ZodTypeDef {
    [metaSymbol]?: Metadata;
  }
  interface ZodType {
    /** @desc Add an example value (before any transformations, can be called multiple times) */
    example(example: this["_input"]): this;
  }
  interface ZodDefault<T extends z.ZodTypeAny> {
    /** @desc Change the default value in the generated Documentation to a label */
    label(label: string): this;
  }
  interface ZodObject<
    T extends z.ZodRawShape,
    UnknownKeys extends z.UnknownKeysParam = z.UnknownKeysParam,
    Catchall extends z.ZodTypeAny = z.ZodTypeAny,
    Output = z.objectOutputType<T, Catchall, UnknownKeys>,
    Input = z.objectInputType<T, Catchall, UnknownKeys>,
  > {
    remap<V extends string, U extends { [P in keyof T]?: V }>(
      mapping: U,
    ): z.ZodPipeline<
      z.ZodEffects<this, FlatObject>, // internal type simplified
      z.ZodObject<Remap<T, U, V> & Intact<T, U>>
    >;
  }
}

const exampleSetter = function (
  this: z.ZodType,
  value: (typeof this)["_input"],
) {
  const copy = cloneSchema(this);
  copy._def[metaSymbol]!.examples.push(value);
  return copy;
};

const labelSetter = function (this: z.ZodDefault<z.ZodTypeAny>, label: string) {
  const copy = cloneSchema(this);
  copy._def[metaSymbol]!.defaultLabel = label;
  return copy;
};

const brandSetter = function (
  this: z.ZodType,
  brand?: string | number | symbol,
) {
  return new z.ZodBranded({
    typeName: z.ZodFirstPartyTypeKind.ZodBranded,
    type: this,
    description: this._def.description,
    errorMap: this._def.errorMap,
    [metaSymbol]: { examples: [], ...clone(this._def[metaSymbol]), brand },
  });
};

const objectMapper = function (
  this: z.ZodObject<z.ZodRawShape>,
  mapping: Record<string, string>,
) {
  return this.transform(
    pipe(
      toPairs,
      map(([key, value]) => pair(mapping[key] || key, value)),
      fromPairs,
    ),
  ).pipe(
    z
      .object(
        pipe(
          toPairs,
          map(([key, schema]) => pair(mapping[String(key)] || key, schema)),
          fromPairs,
        )(clone(this.shape)), // immutable, no references to the original schemas
      )
      [this._def.unknownKeys](), // proxies unknown keys when set to "passthrough"
  );
};

if (!(metaSymbol in globalThis)) {
  (globalThis as Record<symbol, unknown>)[metaSymbol] = true;
  Object.defineProperties(z.ZodType.prototype, {
    ["example" satisfies keyof z.ZodType]: {
      get(): z.ZodType["example"] {
        return exampleSetter.bind(this);
      },
    },
    ["brand" satisfies keyof z.ZodType]: {
      set() {}, // this is required to override the existing method
      get() {
        return brandSetter.bind(this) as z.ZodType["brand"];
      },
    },
  });
  Object.defineProperty(
    z.ZodDefault.prototype,
    "label" satisfies keyof z.ZodDefault<z.ZodTypeAny>,
    {
      get(): z.ZodDefault<z.ZodTypeAny>["label"] {
        return labelSetter.bind(this);
      },
    },
  );
  Object.defineProperty(
    z.ZodObject.prototype,
    "remap" satisfies keyof z.ZodObject<z.ZodRawShape>,
    {
      get() {
        return objectMapper.bind(this) as z.ZodObject<z.ZodRawShape>["remap"];
      },
    },
  );
}
