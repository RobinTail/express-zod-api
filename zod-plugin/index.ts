import "./augmentation";
import "./runtime"; // side effects here
export { pack, unpack } from "./packer";
export { getBrand } from "./brand";
