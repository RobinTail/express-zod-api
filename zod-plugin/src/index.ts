import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
