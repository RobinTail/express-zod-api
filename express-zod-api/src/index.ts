import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
