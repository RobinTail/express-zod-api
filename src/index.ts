import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
