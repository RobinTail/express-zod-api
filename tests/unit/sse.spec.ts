import "../../src/zod-plugin"; // required for this test // @todo until import from index
import { z } from "zod";
import { makeEventSchema } from "../../src/sse";

describe("SSE", () => {
  describe("makeEventSchema()", () => {
    test("should make a valid schema of SSE event", () => {
      expect(makeEventSchema("test", z.string())).toMatchSnapshot();
    });
  });
});
