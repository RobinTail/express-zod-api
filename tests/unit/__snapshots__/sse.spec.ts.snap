// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SSE > makeEventSchema() > should make a valid schema of SSE event 1`] = `
{
  "_type": "ZodObject",
  "shape": {
    "data": {
      "_type": "ZodString",
    },
    "event": {
      "_type": "ZodLiteral",
      "value": "test",
    },
    "id": {
      "_type": "ZodOptional",
      "value": {
        "_type": "ZodString",
      },
    },
    "retry": {
      "_type": "ZodOptional",
      "value": {
        "_type": "ZodNumber",
      },
    },
  },
}
`;

exports[`SSE > makeResultHandler() > should create ResultHandler describing possible events and handling generic errors 1`] = `
[
  {
    "mimeTypes": [
      "text/event-stream",
    ],
    "schema": {
      "_type": "ZodDiscriminatedUnion",
      "discriminator": "event",
      "options": [
        {
          "_type": "ZodObject",
          "shape": {
            "data": {
              "_type": "ZodString",
            },
            "event": {
              "_type": "ZodLiteral",
              "value": "test",
            },
            "id": {
              "_type": "ZodOptional",
              "value": {
                "_type": "ZodString",
              },
            },
            "retry": {
              "_type": "ZodOptional",
              "value": {
                "_type": "ZodNumber",
              },
            },
          },
        },
        {
          "_type": "ZodObject",
          "shape": {
            "data": {
              "_type": "ZodNumber",
            },
            "event": {
              "_type": "ZodLiteral",
              "value": "another",
            },
            "id": {
              "_type": "ZodOptional",
              "value": {
                "_type": "ZodString",
              },
            },
            "retry": {
              "_type": "ZodOptional",
              "value": {
                "_type": "ZodNumber",
              },
            },
          },
        },
      ],
    },
    "statusCodes": [
      200,
    ],
  },
]
`;

exports[`SSE > makeResultHandler() > should create ResultHandler describing possible events and handling generic errors 2`] = `
[
  {
    "mimeTypes": [
      "text/plain",
    ],
    "schema": {
      "_type": "ZodString",
    },
    "statusCodes": [
      400,
    ],
  },
]
`;
